src/main/java/uk/gov/hmcts/reform/sandl/snlevents/actions/hearing/WithdrawHearingAction.java

```diff
@@ -0,0 +1,189 @@
+import uk.gov.hmcts.reform.sandl.snlevents.model.db.Hearing;
+import uk.gov.hmcts.reform.sandl.snlevents.model.db.HearingPart;
+import uk.gov.hmcts.reform.sandl.snlevents.model.db.Session;
+import uk.gov.hmcts.reform.sandl.snlevents.model.db.UserTransactionData;
+import uk.gov.hmcts.reform.sandl.snlevents.model.request.VersionInfo;
+import uk.gov.hmcts.reform.sandl.snlevents.model.request.WithdrawHearingRequest;
+import uk.gov.hmcts.reform.sandl.snlevents.repository.db.HearingPartRepository;
+import uk.gov.hmcts.reform.sandl.snlevents.repository.db.HearingRepository;
+import uk.gov.hmcts.reform.sandl.snlevents.service.RulesService;
+import uk.gov.hmcts.reform.sandl.snlevents.service.StatusConfigService;
+import uk.gov.hmcts.reform.sandl.snlevents.service.StatusServiceManager;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+public class WithdrawHearingAction extends Action implements RulesProcessable {
+
+    protected WithdrawHearingRequest withdrawHearingRequest;
+    protected Hearing hearing;
+    protected List<HearingPart> hearingParts;
+    protected List<Session> sessions;
+
+    protected HearingRepository hearingRepository;
+    protected HearingPartRepository hearingPartRepository;
+    protected StatusConfigService statusConfigService;
+    protected StatusServiceManager statusServiceManager;
+
+    // id & hearing part string
+    private Map<UUID, String> originalHearingParts;
+    private String previousHearing;
+
+    public WithdrawHearingAction(
+        WithdrawHearingRequest withdrawHearingRequest,
+        HearingRepository hearingRepository,
+        HearingPartRepository hearingPartRepository,
+        StatusConfigService statusConfigService,
+        StatusServiceManager statusServiceManager,
+        ObjectMapper objectMapper
+    ) {
+        this.withdrawHearingRequest = withdrawHearingRequest;
+        this.hearingRepository = hearingRepository;
+        this.hearingPartRepository = hearingPartRepository;
+        this.statusConfigService = statusConfigService;
+        this.statusServiceManager = statusServiceManager;
+        this.objectMapper = objectMapper;
+    }
+
+    @Override
+    public void getAndValidateEntities() {
+        hearing = hearingRepository.findOne(withdrawHearingRequest.getHearingId());
+        hearingParts = hearing.getHearingParts();
+        sessions = hearingParts.stream()
+            .map(HearingPart::getSession)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+        // Validation moved to act() due to conflict with optimistic lock
+
+        if (!statusServiceManager.canBeWithdrawn(hearing)) {
+            throw new SnlEventsException("Hearing can not be withdrawn");
+        }
+        hearingParts.forEach(hp -> {
+            if (!statusServiceManager.canBeWithdrawn(hp)) {
```

Comment:
I think that we don't need to check if hearingPart can be withdrawn. In our system we only take actions on Hearing,  the only place where we take action on hearingPart is planner where we can reassign hearingPart but only if it's single session. @kpakur is it right?